File: python/lexer.py
import json
from enum import Enum, auto
from typing import Callable

class TokenKind(Enum):
    UNKNOWN             = 0
    EOL                 = auto()
    TEXT                = auto()
    NUMBER              = auto()
    NEWLINE             = auto()
    COMMA               = auto()
    DOT                 = auto()
    QUOTE               = auto() # single quotes
    EXCLAIMATION        = auto()
    COLON               = auto()
    SEMICOLON           = auto()
    OPEN_PAREN          = auto()
    CLOSE_PAREN         = auto()
    PUNCT               = auto() # anything else defaults to any punctuation

class Context(Enum):
    UNKNOWN             = 0
    ENTRY_WORD          = auto()
    DEFINITION          = auto()
    EXAMPLES            = auto()
    SYNONYM             = auto()

class Token:
    def __init__(self):
        self.kind: TokenKind = TokenKind.UNKNOWN
        self.content: str = ""

    def __str__(self):
        return json.dumps({
            "token": self.kind.name,
            "content": self.content
        })

class Lexer:
    def __init__(self, content: str):
        self.content: str = content
        self.cursor: int = 0

    def _trim_left(self) -> None:
        while (self.cursor != len(self.content)
               and self.content[self.cursor].isspace()):
            self.cursor += 1

    def next(self) -> Token:
        token = Token()
        content: list[str] = []

        self._trim_left()

        def _consume(predicate: Callable[[str], bool], kind: TokenKind) -> None:
            while predicate(self.content[self.cursor]):
                content.append(self.content[self.cursor])
                self.cursor += 1

            token.kind = kind
            token.content = "".join(content)

        if self.cursor == len(self.content):
            token.kind = TokenKind.EOL

        elif self.content[self.cursor].isalpha():
            _consume(lambda x: x.isalpha(), TokenKind.TEXT)

        elif self.content[self.cursor].isnumeric():
            _consume(lambda x: x.isnumeric(), TokenKind.NUMBER)

        else:
            match self.content[self.cursor]:
                case ',':
                    token.kind = TokenKind.COMMA
                case '!':
                    token.kind = TokenKind.EXCLAIMATION
                case '.':
                    token.kind = TokenKind.DOT
                case "'":
                    token.kind = TokenKind.QUOTE
                case ':':
                    token.kind = TokenKind.COLON
                case ';':
                    token.kind = TokenKind.SEMICOLON
                case '(':
                    token.kind = TokenKind.OPEN_PAREN
                case ')':
                    token.kind = TokenKind.CLOSE_PAREN
                case _:
                    token.kind = TokenKind.PUNCT

            token.content = self.content[self.cursor]
            self.cursor += 1

        return token

    def __iter__(self):
        self.cursor = 0
        return self

    def __next__(self):
        token = self.next()
        if token.kind == TokenKind.EOL:
            raise StopIteration

        return token

File: parser/src/lexer.ts
import { isspace, isdigit, isword, isabbv, isalpha } from "./util";

enum TokenKind {
    UNKNOWN,
    EOL,
    TEXT,
    NUMBER,
    NEWLINE,
    COMMA,
    DOT,
    DOLLAR,
    EXCLAIMATION,
    QUESTION,
    COLON,
    SEMICOLON,
    OPEN_PAREN,
    CLOSE_PAREN,
    CARET,
    PUNCT,
}

class Token {
    kind: TokenKind;
    content: string;

    constructor() {
        this.kind = TokenKind.UNKNOWN;
        this.content = "";
    }

    toString(): string {
        return `[ kind: ${TokenKind[this.kind]}, content: "${this.content}" ]`;
    }
}

class Lexer {
    private cursor: number;
    private short: string;
    readonly content: string;
    readonly tokens: Token[];

    constructor(content: string) {
        this.content = content;
        this.cursor = 0;
        this.tokens = [];
        this.short = `${content[0].toLowerCase()}.`;

        this.tokenize();
    }

    /**
     * Removes leading whitespaces.
     */
    private trim_left(): void {
        while (this.cursor != this.content.length
            && isspace(this.content[this.cursor]))
            this.cursor += 1
    }

    /**
     * Returns the current token and iterate to the next.
     */
    private next(): Token {
        let token = new Token();
        let content: string[] = [];

        this.trim_left();

        /**
         * Consumes an entire word.
         * @param ... first two should be obvious
         * @param abbv {boolean} should this function checks for abbv?
         */
        const consume = (predicate: (str: string) => boolean,
                         kind: TokenKind,
                         abbv: boolean = false): void => {
            while (predicate(this.content[this.cursor])) {
                content.push(this.content[this.cursor]);
                this.cursor += 1;
            }

            if (abbv && this.content[this.cursor] == ".") {
                let abbv = [...content];
                let temp = this.cursor;

                while ((isword(this.content[this.cursor])
                        || this.content[this.cursor] == ".")
                       && this.content.length != this.cursor) {
                    abbv.push(this.content[this.cursor]);
                    this.cursor += 1;
                }

                let shorten = abbv.slice(abbv.length - 2, abbv.length)
                                  .join("")
                                  .toLowerCase();

                if (isabbv(abbv.join(""))
                    || (this.short == shorten
                        && !isalpha(abbv[abbv.length - 3]))) {
                    content = abbv;
                } else
                    this.cursor = temp;
            }

            token.kind = kind;
            token.content = content.join("");
        }

        if (this.cursor == this.content.length)
            token.kind = TokenKind.EOL;
        else if (isword(this.content[this.cursor]))
            consume((str) => isword(str), TokenKind.TEXT, true);
        else if (isdigit(this.content[this.cursor]))
            consume((str) => isdigit(str), TokenKind.NUMBER);
        else {
            switch (this.content[this.cursor]) {
                case ".":
                    token.kind = TokenKind.DOT; break;
                case ",":
                    token.kind = TokenKind.COMMA; break;
                case "!":
                    token.kind = TokenKind.EXCLAIMATION; break;
                case "?":
                    token.kind = TokenKind.QUESTION; break;
                case ":":
                    token.kind = TokenKind.COLON; break;
                case ";":
                    token.kind = TokenKind.SEMICOLON; break;
                case "(":
                    token.kind = TokenKind.OPEN_PAREN; break;
                case ")":
                    token.kind = TokenKind.CLOSE_PAREN; break;
                case "^":
                    token.kind = TokenKind.CARET; break;
                case "$":
                    token.kind = TokenKind.DOLLAR; break;
                default:
                    token.kind = TokenKind.PUNCT;
            }

            token.content = this.content[this.cursor];
            this.cursor += 1;
        }

        return token;
    }

    /**
     * Tokenize the string and returns a list of tokens.
     */
    tokenize(): Token[] {
        let token: Token = new Token();

        this.cursor = 0; // reset cursor

        while (token.kind != TokenKind.EOL) {
            if (token.kind != TokenKind.UNKNOWN)
                this.tokens.push(token);
            token = this.next();
        }

        return this.tokens;
    }

    *[Symbol.iterator]() {
        for (let i = 0; i < this.tokens.length; ++i) {
            yield this.tokens[i];
        }
    }
}

export { Lexer, Token, TokenKind };
